packages/fetchai/connections/p2p_libp2p_client/connection.py

```diff
@@ -314,26 +365,34 @@ async def _send(self, data: bytes) -> None:
         self._writer.write(data)
         await self._writer.drain()
 
-    async def _receive(self) -> Optional[bytes]:
+    async def _read_message_from_reader(self) -> Optional[bytes]:
+        """Try to read message from reader."""
         if self._reader is None:
             raise ValueError("Reader is not set.")  # pragma: nocover
+
+        buf = await self._reader.readexactly(4)
+        if not buf:  # pragma: no cover
+            return None
+        size = struct.unpack("!I", buf)[0]
+        data = await self._reader.readexactly(size)
+        if not data:  # pragma: no cover
+            return None
+        return data
+
+    async def _receive(self) -> Optional[bytes]:
+        """Receive binary message."""
         try:
             self.logger.debug("Waiting for messages...")
-            buf = await self._reader.readexactly(4)
-            if not buf:  # pragma: no cover
-                return None
-            size = struct.unpack("!I", buf)[0]
-            data = await self._reader.readexactly(size)
-            if not data:  # pragma: no cover
-                return None
-            return data
+            return await self._read_message_from_reader()
         except ConnectionError as e:  # pragma: nocover
-            self.logger.info(f"Connection error: {e}")
-            return None
+            self.logger.error(f"Connection error: {e}. Try to reconnect and read again")
+            await self._perform_connection_to_node()
+            return await self._read_message_from_reader()
```

Comment:
Will this work if the previous read experienced a disconnection?

Reply:
it sets a completely new connection, so will work